### PR TITLE
crimson: misc fixes for writes to multiple-osd cluster

### PR DESCRIPTION
**Pull request by cyx1231st**
Test scenario:
* 20 crimson-OSDs with cyan store;
* create a replicated test pool with 128 PGs and min-size/size set to 3;
* rados bench 4K random write;

Fixed most of the found issues, including https://github.com/ceph/ceph/pull/33836#discussion_r390745491.

The cluster can almost work now.
Writes may still fail with larger scale tests (50 OSDs, 256 PGs), seems related to peering @tchaikov @xxhdx1985126 @athanatos :
```
INFO  2020-03-14 18:43:37,262 [shard 0] osd - osdmap_subscribe(54)
WARN  2020-03-14 18:43:37,262 [shard 0] monc - renew_subs - empty
INFO  2020-03-14 18:43:37,263 [shard 0] osd - osd.2: now active
WARN  2020-03-14 18:43:37,263 [shard 0] osd - peering_event(id=7, detail=PeeringEvent(from=31 pgid=1.7f sent=103 requested=103 evt=epoch_sent: 103 epoch_requested: 103 MQuery 1.7f from 31 query_epoch 103 query: query(info 0'0 epoch_sent 103))): pg absent, did not create
WARN  2020-03-14 18:43:37,264 [shard 0] osd - peering_event(id=4, detail=PeeringEvent(from=27 pgid=1.d6 sent=103 requested=103 evt=epoch_sent: 103 epoch_requested: 103 MQuery 1.d6 from 27 query_epoch 103 query: query(info 0'0 epoch_sent 103))): pg absent, did not create
terminate called after throwing an instance of 'std::runtime_error'
  what():  read gave enoent on #-1:a502e935:::osdmap.103:0#
Aborting on shard 0.
Backtrace:

auto crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::_future<crimson::errorated_future_marker<ceph::buffer::v15_2_0::list> >::_safe_then_handle_errors<seastar::futurize<seastar::future<ceph::buffer::v15_2_0::list> >, seastar::future<ceph::buffer::v15_2_0::list>, crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}> >(seastar::future<ceph::buffer::v15_2_0::list>&&, crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}>&&) at /root/ceph/src/crimson/common/errorator.h:377
auto crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::_future<crimson::errorated_future_marker<ceph::buffer::v15_2_0::list> >::handle_error<crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}> >(crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}>&&)::{lambda(auto:1)#1}::operator()<seastar::future<ceph::buffer::v15_2_0::list> >(seastar::future<ceph::buffer::v15_2_0::list>) at /root/ceph/src/crimson/common/errorator.h:596
seastar::future<ceph::buffer::v15_2_0::list> seastar::futurize<seastar::future<ceph::buffer::v15_2_0::list> >::apply<crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::_future<crimson::errorated_future_marker<ceph::buffer::v15_2_0::list> >::handle_error<crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}> >(crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}>&&)::{lambda(auto:1)#1}, seastar::future<ceph::buffer::v15_2_0::list> >(crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::_future<crimson::errorated_future_marker<ceph::buffer::v15_2_0::list> >::handle_error<crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}> >(crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}>&&)::{lambda(auto:1)#1}, seastar::future<ceph::buffer::v15_2_0::list>&&) at /root/ceph/src/seastar/include/seastar/core/future.hh:1655
_ZZN7seastar6futureIJN4ceph6buffer7v15_2_04listEEE24then_wrapped_maybe_eraseILb0ES5_ZN7crimson9erroratorIJNS7_19unthrowable_wrapperIRKSt10error_codeL_ZNS7_2ecILSt4errc2EEEEEENS9_ISC_L_ZNSD_ILSE_5EEEEEEEE7_futureINS7_23errorated_future_markerIJS4_EEEE12handle_errorINSH_14all_same_way_tIZN7OSDMeta8load_mapEjEUlvE_EEEEDaOT_EUlSS_E_EENS_8futurizeIT0_E4typeEOT1_ENUlOS5_E_clES10_ at /root/ceph/src/seastar/include/seastar/core/future.hh:1238
seastar::noncopyable_function<seastar::future<ceph::buffer::v15_2_0::list> (seastar::future<ceph::buffer::v15_2_0::list>&&)>::direct_vtable_for<seastar::future<ceph::buffer::v15_2_0::list>::then_wrapped_maybe_erase<false, seastar::future<ceph::buffer::v15_2_0::list>, crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::_future<crimson::errorated_future_marker<ceph::buffer::v15_2_0::list> >::handle_error<crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}> >(crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}>&&)::{lambda(auto:1)#1}>(crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::_future<crimson::errorated_future_marker<ceph::buffer::v15_2_0::list> >::handle_error<crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}> >(crimson::errorator<crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)2> >, crimson::unthrowable_wrapper<std::error_code const&, crimson::ec<(std::errc)5> > >::all_same_way_t<OSDMeta::load_map(unsigned int)::{lambda()#1}>&&)::{lambda(auto:1)#1}&&)::{lambda(seastar::future<ceph::buffer::v15_2_0::list>&&)#1}>::call(seastar::noncopyable_function<seastar::future<ceph::buffer::v15_2_0::list> (seastar::future<ceph::buffer::v15_2_0::list>&&)> const*, seastar::future<ceph::buffer::v15_2_0::list>&&) at /root/ceph/src/seastar/include/seastar/util/noncopyable_function.hh:101
seastar::noncopyable_function<seastar::future<ceph::buffer::v15_2_0::list> (seastar::future<ceph::buffer::v15_2_0::list>&&)>::operator()(seastar::future<ceph::buffer::v15_2_0::list>&&) const at /root/ceph/src/seastar/include/seastar/util/noncopyable_function.hh:184
seastar::future<ceph::buffer::v15_2_0::list> seastar::futurize<seastar::future<ceph::buffer::v15_2_0::list> >::apply<seastar::noncopyable_function<seastar::future<ceph::buffer::v15_2_0::list> (seastar::future<ceph::buffer::v15_2_0::list>&&)>, seastar::future<ceph::buffer::v15_2_0::list> >(seastar::noncopyable_function<seastar::future<ceph::buffer::v15_2_0::list> (seastar::future<ceph::buffer::v15_2_0::list>&&)>&&, seastar::future<ceph::buffer::v15_2_0::list>&&) at /root/ceph/src/seastar/include/seastar/core/future.hh:1655
seastar::future<ceph::buffer::v15_2_0::list>::then_wrapped_common<false, seastar::future<ceph::buffer::v15_2_0::list>, seastar::noncopyable_function<seastar::future<ceph::buffer::v15_2_0::list> (seastar::future<ceph::buffer::v15_2_0::list>&&)> >(seastar::noncopyable_function<seastar::future<ceph::buffer::v15_2_0::list> (seastar::future<ceph::buffer::v15_2_0::list>&&)>&&)::{lambda()#1}::operator()() const::{lambda(seastar::future_state<ceph::buffer::v15_2_0::list>&&)#1}::operator()(seastar::future_state<ceph::buffer::v15_2_0::list>) at /root/ceph/src/seastar/include/seastar/core/future.hh:1265
seastar::continuation<seastar::future<ceph::buffer::v15_2_0::list>::then_wrapped_common<false, seastar::future<ceph::buffer::v15_2_0::list>, seastar::noncopyable_function<seastar::future<ceph::buffer::v15_2_0::list> (seastar::future<ceph::buffer::v15_2_0::list>&&)> >(seastar::noncopyable_function<seastar::future<ceph::buffer::v15_2_0::list> (seastar::future<ceph::buffer::v15_2_0::list>&&)>&&)::{lambda()#1}::operator()() const::{lambda(seastar::future_state<ceph::buffer::v15_2_0::list>&&)#1}, ceph::buffer::v15_2_0::list>::run_and_dispose() at /root/ceph/src/seastar/include/seastar/core/future.hh:506
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /root/ceph/src/seastar/src/core/reactor.cc:2067
seastar::reactor::run_some_tasks() at /root/ceph/src/seastar/src/core/reactor.cc:2482 (discriminator 2)
seastar::reactor::run() at /root/ceph/src/seastar/src/core/reactor.cc:2642
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at /root/ceph/src/seastar/src/core/app-template.cc:199 (discriminator 1)
main at /root/ceph/src/crimson/osd/main.cc:148 (discriminator 1)
__libc_start_main at /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:310
_start at ??:?
```

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
@xxhdx1985126 could you take a look if the problem is gone?

**Comment by xxhdx1985126** [comment]
@cyx1231st will do that:-)

**Comment by cyx1231st** [comment]
found another issue, working on it.
```
crimson-osd: /root/ceph/src/crimson/net/Protocol.cc:48: void crimson::net::Protocol::close(bool, std::optional<std::function<void()> >): Assertion `close_ready.valid()' failed.
```

**Comment by xxhdx1985126** [comment]
@cyx1231st Sorry that I didn't manage to fully test your idea , now that I've fixed those bugs of alienstore that prevented me to do so, I'll do the full test to see if there's other problem:-)

**Comment by xxhdx1985126** [comment]
@cyx1231st did the problem you mentioned in above happened during the osd's first boot? Thanks:-)

**Comment by cyx1231st** [comment]
> did the problem you mentioned in above happened during the osd's first boot? Thanks:-)

Mostly happened when the cluster is large enough to cause connection racing and resets, and/or during replicated writes like after 2 seconds.

**Comment by cyx1231st** [comment]
@tchaikov @xxhdx1985126 

Did more tests, the heartbeat OSD peers are now trapped in an infinite loop of `remove_peer()` and `add_peer()` when lossy connections race against each other. And I think this is the reason why heartbeat in classic OSD creates 2 messengers as client and server.

The infinite loop is mostly caused by:
- The lossy policy always decides that the ACCEPTING side is the winner, close/reset the CONNECTING one; The same thing happens at peer OSD symmetrically, which in turn causes the winner connection be closed/reset later.
- The issue is exaggerated, because when a reset happens, OSD closes both hb_front and hb_back connections. For example even a hb_front connection can be established occasionally, it will be closed by hb_back racing and causes reset.

So the `lossy_client ` policy is also not appropriate with the current heartbeat implementations. And a simple fix would still be to add two hb_client messengers just like classic OSD.

**Comment by xxhdx1985126** [comment]
@cyx1231st Acknowledged

File: src/crimson/net/Protocol.cc [hunk 1]
```diff
@@ -39,37 +40,60 @@ bool Protocol::is_connected() const
...
   }
-
   closed = true;
   set_write_state(write_state_t::drop);
+  auto gate_closed = pending_dispatch.close();
 
-  return close_ready.get_future();
+  // asynchronous operations
+  close_ready = seastar::when_all_succeed(
+    std::move(gate_closed).finally([this] {
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/net/Protocol.cc]: It's to distinguish the atomic part and the async part of `pending_dispatch.close()` operation.
- The messenger requires the gate close immediately upon connection closing.
- It is asynchronous to wait for all the gated continuations to finish.

File: src/crimson/net/Protocol.cc [hunk 2]
```diff
@@ -39,37 +40,60 @@ bool Protocol::is_connected() const
...
+      if (socket) {
+        return socket->close();
+      }
+      return seastar::now();
+    }),
+    [this, dispatch_reset] {
+      if (dispatch_reset) {
+        // force ms_handle_reset() to be an asynchronous task to prevent
+        // internal state contamination.
+        return seastar::sleep(0s).then([this] {
```
Comment (tchaikov) [inline src/crimson/net/Protocol.cc]: i don't quite understand this part. i think you assume that by inserting a bubble, we can force the reactor to schedule `socket->close()` **before** `dispatcher.ms_handle_reset()`, right?

**Comment by tchaikov** [comment]
jenkins test make check

File: src/crimson/net/Protocol.cc (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/net/Protocol.cc]: Yes, this is to force `ms_handle_reset()` to return a not-available future, so it won't be optimized by seastar as part of the current continuation which potentially can pollute the internal states, and it becomes an independent continuation that would be dispatched after `socket->close()`. But later I found out that it causes other issues.

File: src/crimson/osd/heartbeat.cc [hunk 3]
```diff
@@ -287,46 +287,41 @@ seastar::future<> Heartbeat::handle_you_died()
...
-        now + std::chrono::seconds(local_conf()->osd_heartbeat_grace);
-      auto& info = item.second;
-      info.last_tx = now;
-      if (clock::is_zero(info.first_tx)) {
-        info.first_tx = now;
+  std::vector<seastar::future<>> futures;
+  for (auto& item : peers) {
+    const auto mnow = service.get_mnow();
+    const auto now = clock::now();
+    const auto deadline =
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/osd/heartbeat.cc]: Not sure, but I remembered a guideline saying that it's better/safer to declare the variable as close as its first usage. 

Also checked about related recommendations: https://stackoverflow.com/questions/7959573/declaring-variables-inside-loops-good-practice-or-bad-practice 

What do you think?

File: src/crimson/osd/heartbeat.cc [hunk 4]
```diff
@@ -287,46 +287,41 @@ seastar::future<> Heartbeat::handle_you_died()
...
+      now + std::chrono::seconds(local_conf()->osd_heartbeat_grace);
+    auto& info = item.second;
+    info.last_tx = now;
+    if (clock::is_zero(info.first_tx)) {
+      info.first_tx = now;
+    }
+    const utime_t sent_stamp{now};
+    [[maybe_unused]] auto [reply, added] =
+      info.ping_history.emplace(sent_stamp, reply_t{deadline, 0});
+    std::vector<crimson::net::ConnectionRef> conns{info.con_front,
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/osd/heartbeat.cc]: Thanks for the better options, fixed.

File: src/crimson/osd/heartbeat.cc [hunk 5]
```diff
@@ -287,46 +287,41 @@ seastar::future<> Heartbeat::handle_you_died()
...
+          monc.get_fsid(),
+          service.get_osdmap_service().get_map()->get_epoch(),
+          MOSDPing::PING,
+          sent_stamp,
+          mnow,
+          mnow,
+          service.get_osdmap_service().get_up_epoch(),
+          min_message);
+        reply->second.unacknowledged++;
+        futures.push_back(std::move(con->send(ping)));
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/osd/heartbeat.cc]: done.

File: src/crimson/osd/heartbeat.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/osd/heartbeat.cc]: i am not sure if the guideline or the recommendations covers the constant case. instead, i think it's better to be explicit that these are constants.

[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/osd/heartbeat.cc]: Ah, lemme change back...

[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/osd/heartbeat.cc]: reverted

[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/osd/heartbeat.cc]: @cyx1231st i still think we could save some calls of `now()` , as it's very likely implemented using `gettimeofday()` even in modern linux, vDSO is implemented using this function. but it's still involves an extra function call in userspace in each iteration of this for loop. but i agree with you it's always desirable to move the variable definition closer where it's used. but IMHO, they are already visually close..

**Comment by xxhdx1985126** [comment]
Um, hi, I see that there's a commit in this pr that change heartbeat msgr policy to lossy_client, has the issues here https://github.com/ceph/ceph/pull/33909#issuecomment-599380821 has been solved?

File: src/crimson/osd/heartbeat.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/osd/heartbeat.cc]: I guess you mean to use `seastar::lowres_clock::now()` and `seastar::lowres_system_clock::now()` instead, right? I agree, might need to grep our code to replace the slower `now()` altogether.

**Comment by cyx1231st** [comment]
@xxhdx1985126 not yet. I dropped the `lossy_client` policy change in this PR for its incompleteness.

File: src/crimson/osd/heartbeat.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/osd/heartbeat.cc]: in this case, i am just suggesting pulling the constant out of the for loop. we can audit the code to use seastar's counterparts later.

[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/osd/heartbeat.cc]: Moved out of the for loop.

**Pull request merged**